### DISCUSSION TITLE
Ideal behavior of the `//` operator?

### DISCUSSION THREAD

**Discussion by eitsupi** [duckdb/duckdb#7558] [Q&A]
Currently, the new `//` operator seems matching the `/` operator of Postgres and SQLite; which do not return an integer for calculations that include float types.

```sql
DuckDB Web Shell
Database: v0.7.2-dev3783
Package:  @duckdb/duckdb-wasm@1.25.1-dev13.0

Connected to a local transient in-memory database.
Enter .help for usage hints.

duckdb> select 13 // 5;
┌───────────┐
│ (13 // 5) │
╞═══════════╡
│         2 │
└───────────┘
Elapsed: 8 ms

duckdb> select 13 // 5.0;
┌─────────────┐
│ (13 // 5.0) │
╞═════════════╡
│         2.6 │
└─────────────┘
Elapsed: 7 ms

duckdb> select 13.0 // 5;
┌─────────────┐
│ (13.0 // 5) │
╞═════════════╡
│         2.6 │
└─────────────┘
Elapsed: 2 ms

duckdb> select 13.0 // 5.0;
┌───────────────┐
│ (13.0 // 5.0) │
╞═══════════════╡
│           2.6 │
└───────────────┘
Elapsed: 28 ms
```

I think this behavior is different from Python’s `//` operator.
Python’s `//` operator returns a float when the operands are floats, but truncates the decimal part.

```python
In [1]: 13 // 5
Out[1]: 2

In [2]: 13.0 // 5
Out[2]: 2.0

In [3]: 13 // 5.0
Out[3]: 2.0
```

Also, I think the `DIV` operator of MySQL, which has a similar function, always returns an integer quotient.

I wonder if changes will be made in the future to provide exactly the same behavior as Python's `//` operator or MySQL's `DIV` operator.

**Comment by eitsupi** [comment]
Maybe related to #2981

**Reply by paultiq** [reply]
Interesting, for what it's worth, from reading the PR, I thought it was intended to mirror Python, based on: https://github.com/duckdb/duckdb/pull/7082

PEP238 explains Python's implementation:
https://peps.python.org/pep-0238/

**Reply by eitsupi** [reply]
Yeah, another big difference is that Python's `//` is floor devision.

```python
>>> 13 // -5
-3
```

```sql
v0.8.0 e8e4cea5ec
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D select 13 // -5;
┌────────────┐
│ (13 // -5) │
│   int32    │
├────────────┤
│         -2 │
└────────────┘
```

In DuckDB, it is needed to using the `FDIV` macro for now.

```sql
D select FDIV (13, -5);
┌──────────────┐
│ fdiv(13, -5) │
│    double    │
├──────────────┤
│         -3.0 │
└──────────────┘
```
